### PR TITLE
Add CartableOptionPointer and helpers to the yoke crate

### PR DESCRIPTION
**Pull request by sffc**
Split from #4442

I do think this is in scope for `yoke` because the type can be a "one-way" type that is a more efficient cart than either `Option<Rc<T>>` or `Option<ErasedRcCart>`.

### CODE REVIEW

**Review by Manishearth — CHANGES_REQUESTED** [review]
I'm opposed to this, for a bunch of reasons

Firstly, I think the use case here is rather specific; I don't think it's something that makes a lot of sense for most yoke users overall. If we get asks for something like this I'd like to design it based on the asks rather than genericizing the thing that ICU4X needs in a particular direction and assuming that that direction is what is needed.


Secondly, the abstraction itself is incredibly generic in a way that is both rather hard to understand  for users (IMO) and rather hard to grapple with the safety of. I think we're introducing a lot of concepts here that interact in subtle ways.

(For this to land it is going to need a _lot_ more careful documentation of the safety needs)

**Comment by sffc** [comment]
> Firstly, I think the use case here is rather specific; I don't think it's something that makes a lot of sense for most yoke users overall. If we get asks for something like this I'd like to design it based on the asks rather than genericizing the thing that ICU4X needs in a particular direction and assuming that that direction is what is needed.

I don't agree that it is ICU4X-specific. This new cart type `OptionCartablePointer` is another in-house cart abstraction, like `ErasedRcCart`, which is better than any of the other cart abstractions being offered. This is especially true because we already have a lot of code specializing over both `Rc` and `Option` carts, so we clearly have already put a stake in the ground that we expect `Option<Rc>` to be a cart we expect people to want when they are using Yoke.

> Secondly, the abstraction itself is incredibly generic in a way that is both rather hard to understand for users (IMO) and rather hard to grapple with the safety of. I think we're introducing a lot of concepts here that interact in subtle ways.

Almost all of the unsafety is encapsulated in the new `CartablePointerLike` trait. I do not think it is too generic. In fact it turns out that there actually are not very many safety invariants at play; at the end of the day this type is just an efficient way to store the "raw" representation of pointer-like types (reference, Box, Rc, and Arc).

> (For this to land it is going to need a _lot_ more careful documentation of the safety needs)

Would like to pair with you today on writing the documentation.

File: utils/yoke/src/cartable_ptr.rs [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            unsafe { C::clone_raw(ptr) }
+        }
+        Self {
+            inner: self.inner,
+            _cartable: PhantomData,
+        }
+    }
+}
+
+// Safety: logically an Option<C>. Has same bounds as Option<C>
```
Comment (Manishearth) [inline utils/yoke/src/cartable_ptr.rs:312]: nit:  with my other PR landing you can potentially relate to the invariants specified there

**Review by Manishearth — COMMENTED** [review]
code looks good, minor issues around docs and examples

**Review by Manishearth — APPROVED** [review]

**Pull request merged**